### DISCUSSION TITLE
Support case-insensitive comparisons

### DISCUSSION THREAD

**Discussion by chris-martin** [typeclasses/ascii#9] [Ideas]
Previously brought up by @frasertweedale [here](https://github.com/snoyberg/ascii/issues/6#issuecomment-624384806), where the request was to add a `FoldCase` instance (I presume from the [case-insensitive](https://hackage.haskell.org/package/case-insensitive) package).

I'm inclined to discourage use of the `CI` type; it bothers me a lot that `CI` has a field that `==` doesn't consider in its equality test, I think it's an abuse of `Eq`. As the docs say:

> `(==)` is customarily expected to implement an equivalence relationship where two values comparing equal are indistinguishable by "public" functions

And I think that going against this expectation can lead us to unhappy places.

I was considering adding another package called `ascii-subset`, whose purpose would be to provide types like "case-insensitive character". (I can think of a handful of ways to write this: to define a new type for each ASCII subset we care to consider, to define a 128-constructor GADT with some phantom type parameters that allow you to narrow the type, or to define various smaller types like `ControlCharacter`, `Digit`, `CapitalLetter`, `SmallLetter` etc. and then build up various ASCII subsets out of those. But, setting aside these details for the moment...) What do you think, would having "case-insensitive char" and "case-insensitive string" types be helpful, or is there a particular need for the `CI` approach that preserves the original case?

**Comment by frasertweedale** [comment]
In my use case - email and in particular header and parameter names - case-insensitive comparison is essential, and case preservation of values to be compared case-insensitively is desirable.  The *case-insensitive* library meets these requirements.

I take your point about `(==)` and observational equivalence, but what `CI` does it pretty clear and unsurprising.  And useful.  Perhaps implement `FoldCase` with a prominent admonition about `CI`'s violation of the `Eq` substitutivity law and suggest to write ``(==) `on` foldedCase`` instead?

**Comment by frasertweedale** [comment]
And y'know... I'll just leave this here :)
```
λ> import Data.ByteString.Lazy as L
λ> l1 = L.fromChunks ["ab", "c"]
λ> l2 = L.fromChunks ["a", "bc"]
λ> f = L.toChunks
λ> (l1 == l2, f l1 == f l2)
(True,False)
```

**Comment by chris-martin** [comment]
Heh, fair point about lazy ByteString, though I don't have to like that either!

I suppose it's no big deal to add an instance. That dependency is pretty stable, in Stackage, no major transitive dependencies, and the `FoldCase` class has remained unchanged for nine years, so the bound would be `case-insensitive >= 0.1 && < 1.3`.

I do also want to add some [Equivalence](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Functor-Contravariant.html#t:Equivalence)s, but we can do both.
